### PR TITLE
[SPARK-33241][SQL] Dynamic pruning on data column

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr add support dynamic pruning on data column.  Dynamic pruning on data column support 2 cases:
1. Pruning on aggregate function to reduce shuffle data:
image

2. Pruning on sort merge join to reduce shuffle data:
image

This PR introduces a new configuration: `spark.sql.optimizer.dynamicFilterPruning.enabled,` which is the master switch of Dynamic Filter Pruning, `spark.sql.optimizer.dynamicPartitionPruning.enabled` is the switch of the partition column, `spark.sql.optimizer.dynamicDataPruning.enabled` is the switch of the data column.

In addition, the dynamic pruning on data column also supports pushing down the dynamic filter to the data source.

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and performance test:

1. TPC-DS Performance Evaluation

SQL | Shuffle stage origin data size | Pruning shuffle stage data size | Shuffle stage origin records | Pruning shuffle stage records | Disable shuffle pruning (second) | Enable shuffle pruning (second)
-- | -- | -- | -- | -- | -- | --
q13 | 3.2 GiB | 93.1 MiB | 86,409,332 | 1,480,662 | 13 | 14
q16 | 158.2 GiB | 270.7 MiB | 7,136,969,739 | 10,295,246 | 84 | 36
q24a | 355.6 GiB | 39.7 GiB | 13,428,037,922 | 1,504,810,137 | 660 | 432
q24b | 355.6 GiB | 39.7 GiB | 13,428,037,922 | 1,504,810,137 | 660 | 492
q65 | 37.2 GiB | 37.2 GiB | 2,627,543,089 | 2,627,543,089 | 45 | 45
q72 | 40.9 GiB | 1543.3 MiB | 1,418,327,817 | 47,248,271 | 276 | 38
q80 | 8.3 GiB | 8.2 GiB | 295,853,928 | 292,353,065 | 37 | 36
q85 | 12.8 GiB | 1508.2 MiB | 329,635,219 | 37,337,231 | 16 | 12
q93 | 26.7 GiB | 435.9 MiB | 1,389,592,792 | 20,744,184 | 270 | 258
q94 | 87.6 GiB | 1012.0 MiB | 3,598,433,079 | 34,538,210 | 58 | 27
q95 | 640.5 GiB | 344.6 GiB | 872,596,314 | 793,654,526 | 414 | 402

2. Production Performance Evaluation

Before this PR | After this PR
-- | --
image | image

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -245,6 +245,13 @@ object SQLConf {
     .stringConf
     .createOptional
 
+  val DYNAMIC_FILTER_PRUNING_ENABLED =
+    buildConf("spark.sql.optimizer.dynamicFilterPruning.enabled")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:249]: `DYNAMIC_FILTER_PRUNING_ENABLED` -> `DYNAMIC_COLUMN_PRUNING_ENABLED`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -245,6 +245,13 @@ object SQLConf {
...
+      .doc("When true, we will generate predicate for partition column when it's used as join key")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:250]: `partition column` -> `data column`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:249]: Ur, BTW, is this used in this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -282,6 +289,23 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val DYNAMIC_DATA_PRUNING_ENABLED =
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:292]: This looks like the real config in this PR.
`DYNAMIC_DATA_PRUNING_ENABLED` -> `DYNAMIC_COLUMN_PRUNING_ENABLED`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -282,6 +289,23 @@ object SQLConf {
...
+    buildConf("spark.sql.optimizer.dynamicDataPruning.enabled")
+      .doc("When true, we will generate predicate for column when it's used as join key " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:294]: `column` -> `data column` if this PR isn't aiming for `partition column`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:249]: Yes. This PR introduces a new configuration: `spark.sql.optimizer.dynamicFilterPruning.enabled`, which is the master switch of Dynamic Filter Pruning, `spark.sql.optimizer.dynamicPartitionPruning.enabled` is the switch of the partition column, `spark.sql.optimizer.dynamicDataPruning.enabled` is the switch of the data column.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:292]: This config is the switch of is the switch of the data column. May be `DYNAMIC_DATA_COLUMN_PRUNING_ENABLED`?
But this is inconsistent with `DYNAMIC_PARTITION_PRUNING_ENABLED`.

**Pull request closed without merging**